CMakeLists.txt

```diff
@@ -113,84 +111,73 @@ if(HAVE_BUILTIN_EXTENSION_ZSTD)
-# directories.
-set_target_properties(wiredtiger PROPERTIES NO_SYSTEM_FROM_IMPORTED TRUE)
-
-if(ENABLE_STATIC AND NOT ENABLE_PYTHON)
-    # Avoid compiling with fPIC for static builds, this can often lead to
-    # GCC disabling many optimizations e.g. inlining, possibly introducing
-    # an overhead. We additionally can't turn off fPIC if we are compiling
-    # with python support as the python API library depends on shared library
-    # support.
-    set_property(TARGET wiredtiger PROPERTY POSITION_INDEPENDENT_CODE OFF)
-else()
-    set_property(TARGET wiredtiger PROPERTY POSITION_INDEPENDENT_CODE ON)
-endif()
 
-# Ensure we link any available library dependencies to our wiredtiger target.
-if(HAVE_LIBPTHREAD)
-    target_link_libraries(wiredtiger PUBLIC ${HAVE_LIBPTHREAD})
-    if(HAVE_LIBPTHREAD_INCLUDES)
-        target_include_directories(wiredtiger PUBLIC ${HAVE_LIBPTHREAD_INCLUDES})
+# If compiling with fPIC, we can create a intermediate library of
+# position independent objects, that both the static and shared builds
+# can use. This saving on the cost of recompiling the WiredTiger sources
+# two times over.
+if(WITH_PIC OR ENABLE_SHARED)
+    add_library(wt_objs OBJECT ${wt_sources})
+    target_include_directories(wt_objs
+        PRIVATE
+        ${CMAKE_CURRENT_BINARY_DIR}/include
+        ${CMAKE_CURRENT_BINARY_DIR}/config
+        ${CMAKE_CURRENT_LIST_DIR}/src/include
+    )
+    if(ENABLE_TCMALLOC)
+        target_include_directories(wt_objs PRIVATE ${HAVE_LIBTCMALLOC_INCLUDES})
     endif()
-endif()
-
-if(HAVE_LIBRT)
-    target_link_libraries(wiredtiger PUBLIC ${HAVE_LIBRT})
-    if(HAVE_LIBRT_INCLUDES)
-        target_include_directories(wiredtiger PUBLIC ${HAVE_LIBRT_INCLUDES})
+    if(HAVE_LIBPTHREAD)
+        target_include_directories(wt_objs PRIVATE ${HAVE_LIBPTHREAD_INCLUDES})
     endif()
-endif()
-
-if(HAVE_LIBDL)
-    target_link_libraries(wiredtiger PUBLIC ${HAVE_LIBDL})
-    if(HAVE_LIBDL_INCLUDES)
-        target_include_directories(wiredtiger PUBLIC ${HAVE_LIBDL_INCLUDES})
+    if(HAVE_LIBRT)
+        target_include_directories(wt_objs PRIVATE ${HAVE_LIBRT_INCLUDES})
     endif()
+    if(HAVE_LIBDL)
+        target_include_directories(wt_objs PRIVATE ${HAVE_LIBDL_INCLUDES})
+    endif()
+    set_property(TARGET wt_objs PROPERTY POSITION_INDEPENDENT_CODE ON)
 endif()
 
-if(ENABLE_TCMALLOC)
-    target_link_libraries(wiredtiger PRIVATE wt::tcmalloc)
-endif()
-
-# We want to capture any transitive dependencies associated with the builtin library
-# target and ensure we are explicitly linking the 3rd party libraries.
-if(HAVE_BUILTIN_EXTENSION_LZ4)
-    target_link_libraries(wiredtiger PRIVATE wt::lz4)
-endif()
-
-if(HAVE_BUILTIN_EXTENSION_SNAPPY)
-    target_link_libraries(wiredtiger PRIVATE wt::snappy)
-endif()
-
-if(HAVE_BUILTIN_EXTENSION_SODIUM)
-    target_link_libraries(wiredtiger PRIVATE wt::sodium)
+if(ENABLE_STATIC)
+    if(WITH_PIC)
+        define_wiredtiger_library(wiredtiger_static STATIC
```

Comment:
We don't need to call `set_property(TARGET wt_objs PROPERTY POSITION_INDEPENDENT_CODE ON)` here since `WITH_PIC` is true?

Reply:
Good question, that's not needed since we do it here:
https://github.com/wiredtiger/wiredtiger/blob/7690440c8e94dda24d5f0ff7ee886bfc35bafc39/CMakeLists.txt#L143
If `WITH_PIC` is true, the PIC property would already be set in the earlier statement.

Reply:
Oh yes, my bad. I misread the logic, thanks!